pkg/deployer/lib/extension/hook.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package extension
+
+import (
+	"fmt"
+)
+
+type HookType string
+
+const (
+	StartHook                     HookType = "Start"                     // called at the beginning of reconciliation
+	DuringResponsibilityCheckHook HookType = "DuringResponsibilityCheck" // called after the responsibility has been checked but before the reconcile is aborted, in case the deployer is not responsible
+	AfterResponsibilityCheckHook  HookType = "AfterResponsibilityCheck"  // called after it has been determined that the deployer is responsible for the deploy item
+	ShouldReconcileHook           HookType = "ShouldReconcile"           // called when determining whether to reconcile the deploy item
+	BeforeAbortHook               HookType = "BeforeAbort"               // called before aborting a deploy item
+	BeforeForceReconcileHook      HookType = "BeforeForceReconcile"      // called before performing a forced reconcile
+	BeforeDeleteHook              HookType = "BeforeDelete"              // called before deleting a deploy item
+	BeforeReconcileHook           HookType = "BeforeReconcile"           // called before a default reconcile
+	BeforeAnyReconcileHook        HookType = "BeforeAnyReconcile"        // a combination of BeforeAbort, BeforeForceReconcile, BeforeDelete, and BeforeReconcile
+	EndHook                       HookType = "End"                       // called at the end of reconciliation
+)
+
+type ReconcileExtensionHook interface {
+	// Name returns the name of the extension hook.
+	// It is mainly used for logging purposes, but names are expected to be unique.
+	Name() string
+
+	// Type returns the hook type of the extension hook.
+	Type() HookType
```

Comment:
maybe type should be a number so that we could register multiple hooks like `StartHook|EndHook`.
As it is done in the fs.Mod package

Reply:
I have refactored the extension hooks and it is now possibly to register a function for multiple hook types at once, which was the main point of your proposal.

I am not a fan of bit arithmetics because it is unintuitive and more difficult to understand and parse. I don't see any advantages in switching to numbers, except for a probably insignificant improvement in memory consumption, therefore I'd prefer to stick to strings. Feel free to discuss if you see additional advantages.